cocotb/share/lib/fli/FliImpl.h

```diff
@@ -150,102 +150,75 @@ class FliTimedCbHdl : public FliProcessCbHdl {
-protected:
-    int m_acc_type;
-    int m_acc_full_type;
-};
-
-class FliObjHdl : public GpiObjHdl, public FliObj {
+// Object Handles
+class FliObjHdl : public GpiObjHdl {
 public:
     FliObjHdl(GpiImplInterface *impl,
               void *hdl,
-              gpi_objtype_t objtype,
-              int acc_type,
-              int acc_full_type) :
-                  GpiObjHdl(impl, hdl, objtype, false),
-                  FliObj(acc_type, acc_full_type) { }
-
-    FliObjHdl(GpiImplInterface *impl,
-              void *hdl,
-              gpi_objtype_t objtype,
-              int acc_type,
-              int acc_full_type,
-              bool is_const) :
-                  GpiObjHdl(impl, hdl, objtype, is_const),
-                  FliObj(acc_type, acc_full_type) { }
+              gpi_objtype_t objtype) :
+                  GpiObjHdl(impl, hdl, objtype, false) { }
 
     virtual ~FliObjHdl() { }
 
-    virtual int initialise(std::string &name, std::string &fq_name);
+    virtual int initialise(std::string &name, std::string &fq_name, fli_type_t fli_type);
 };
 
-class FliSignalObjHdl : public GpiSignalObjHdl, public FliObj {
+class FliSignalObjHdl : public GpiSignalObjHdl {
 public:
     FliSignalObjHdl(GpiImplInterface *impl,
                     void *hdl,
                     gpi_objtype_t objtype,
-                    bool is_const,
-                    int acc_type,
-                    int acc_full_type,
-                    bool is_var) :
+                    bool is_const) :
                         GpiSignalObjHdl(impl, hdl, objtype, is_const),
-                        FliObj(acc_type, acc_full_type),
-                        m_is_var(is_var),
-                        m_rising_cb(impl, this, GPI_RISING),
-                        m_falling_cb(impl, this, GPI_FALLING),
-                        m_either_cb(impl, this, GPI_FALLING | GPI_RISING) { }
-
-    virtual ~FliSignalObjHdl() { }
+                        m_rising_cb(NULL),
+                        m_falling_cb(NULL),
+                        m_either_cb(NULL) { }
+
+    virtual ~FliSignalObjHdl() {
+        if (m_rising_cb)
+            delete m_rising_cb;
+        if (m_falling_cb)
+            delete m_falling_cb;
+        if (m_either_cb)
+            delete m_either_cb;
+    }
 
     virtual GpiCbHdl *value_change_cb(unsigned int edge);
-    virtual int initialise(std::string &name, std::string &fq_name);
-
-    bool is_var(void) { return m_is_var; }
+    virtual int initialise(std::string &name, std::string &fq_name, fli_type_t fli_type);
 
 protected:
-    bool               m_is_var;
-    FliSignalCbHdl     m_rising_cb;
-    FliSignalCbHdl     m_falling_cb;
-    FliSignalCbHdl     m_either_cb;
+    FliSignalCbHdl      *m_rising_cb;
+    FliSignalCbHdl      *m_falling_cb;
+    FliSignalCbHdl      *m_either_cb;
+    mti_GetValue         m_get_value;
+    mti_SetValue         m_set_value;
+    mti_GetArrayValue    m_get_array_value;
+    mti_GetValueIndirect m_get_value_indirect;
```

Comment:
This looks like a use case for virtual functions, not function pointers and if statements

Reply:
Actually, I disagree with that assessment.  FLI is a pain purely because it treats Regions, Signal and Variables differently.  Signals and Variables are essentially the same, just different handle type and function name.  If I went the virtual function approach there would be twice as many classes that are derived from FliValueObjHdl because you would have to have a sub-class for signals and a sub-class for variables.  The code in these classes would be identical to each other, except the name of the function be called.  That seems to open up the potential of bugs getting fixed in one and not the other. 

The function pointer approach allows the classes to not care whether it is operating on a signal or a variable, which allows for a single common code base to be maintained.

Reply:
> The code in these classes would be identical to each other, except the name of the function be called. That seems to open up the potential of bugs getting fixed in one and not the other.

You'd make them share a base class. The idea is that instead of storing the FLI type in a variable, you store it in the type of the object itself. That way you don't have to put if statements in all the methods to switch between types.

Reply:
What if statements are you referring to?   With my approach I have removed all the if statements.  And I don't store the FLI type.